File: old/numerical/romberg.cc
#include<iostream>
#include <cmath>

using namespace std;

typedef double (*FuncOneVar)(double);
class Romberg{ 
private:
  FuncOneVar f;
public:
  Romberg(FuncOneVar f): f(f) {}
  
  double integral(double a, double b) {
    // hard coded but readable
    // in reality, we are adding a series of f(h/2) 
    double h = (b-a);
    double s1 = 0.5 * (f(a) + f(b));
    double s2 = s1 + f(0.5*h);
    double s4 = s2 + f(0.25*h) + f(0.75*h);
    double s8 = s4 + f(0.125*h) + f(0.375*h) + f(0.625*h) + f(0.875*h);
    
    // this shows what h is at each level, h = 1 -. 0.5 -> 0.25, ...
    // h is the distance between the bins normalized over the interval
    // so the whole interval, half of it, then every quarter
    // represents the width of the bins 
    double i1 = s1 * (h*1);
    double i2 = s2 * (h*0.5);
    double i4 = s4 * (h*0.25);
    double i8 = s8 * (h*0.125);
    
    // the error of i2 is 1/4 the error of i1, same relationship between i4 and i2, etc...
    // why that works? math   ask romberg
    double r1 = (4*i2 - i1) / 3;
    double r2 = (4*i4 - i2) / 3;
    double q1 = (16*r2 - r1) / 15;

    return q1;
  }
    
};

double function(double x) {
    return pow(x, 2);
}

int main() {

    Romberg r(function);
    cout << r.integral(0, 1) << endl;
    cout << r.integral(0, 2) << endl;
    cout << r.integral(0, 3) << endl;
    cout << r.integral(0, 4) << endl;
    cout << r.integral(0, 5) << endl;
    cout << r.integral(0, 6) << endl;
}

File: old/numerical/Romberg.java
public class Romberg {
    public static double integrate(FuncOneVar func, double a, double b, double eps) {
        double h, h2, h4, h8;
        double S1, S2, S4, S8;
        double I1, I2, I4, I8;
        double R1, R2, R4;
        double Q1, Q2;
        double Romberg;
        S1 = 0.5*(func.f(a) + func.f(b);
        h = (b-a);
        do {
            h2 = h * 0.5;
            S2 = S1;
            for (double x = a + h2; x < b; x += h)
                S2 += func.f(x);
            h4 = h2 * 0.5;
            S4 = S2;
            for (double x = a + h4; x < b; x += h2)
                S4 += func.f(x);
            h8 = h4 * 0.5;
            S8 = S4;
            for (double x = a + h8; x < b; x += h2)
                S8 += func.f(x);
            I1 = S1 * h;
            I2 = S2 * h2;
            I4 = S4 * h4;
            I8 = S8 * h8;
            R1 = (4*I2 - I1) / 3;
            R2 = (4*I4 - I2) / 3;
            R4 = (4*I8 - I4) / 3;
            Q1 = (16*R2 - R1) / 15;
            Q2 = (16*R4 - R2) / 15;
            Romberg = (64*Q2 - Q1) / 63; // accurate to O(h^8)!!!
            h = h8;
            S1 = S8;
        } while (Math.abs(Q2-Q1) > eps);
        return Romberg;
            
    }

}

interface FuncOneVar {
    public double f(double x);
}

class F1 implements FuncOneVar {
    public double f(double x) { return x*x; }
}